code/controllers/subsystem/security_level.dm

```diff
@@ -1,17 +1,109 @@
+	current_security_level = available_levels[number_level_to_text(SEC_LEVEL_GREEN)]
+
+/datum/controller/subsystem/security_level/fire(resumed)
+	if(!current_security_level.looping_sound) // No sound? No play.
+		can_fire = FALSE
+		return
+	sound_to_playing_players(current_security_level.looping_sound)
+
 
 /**
  * Sets a new security level as our current level
  *
+ * This is how everything should change the security level.
+ *
  * Arguments:
  * * new_level The new security level that will become our current level
  */
 /datum/controller/subsystem/security_level/proc/set_level(new_level)
-	SSsecurity_level.current_level = new_level
-	SEND_SIGNAL(src, COMSIG_SECURITY_LEVEL_CHANGED, new_level)
+	new_level = istext(new_level) ? new_level : number_level_to_text(new_level)
+	if(new_level == current_security_level.name) // If we are already at the desired level, do nothing
+		return
+
+	var/datum/security_level/selected_level = available_levels[new_level]
+
+	if(!selected_level)
+		CRASH("set_level was called with an invalid security level([new_level])")
+
+	announce_security_level(selected_level) // We want to announce BEFORE updating to the new level
+
+	var/old_shuttle_call_time_mod = current_security_level.shuttle_call_time_mod // Need this before we set the new one
+
+	SSsecurity_level.current_security_level = selected_level
+
+	if(selected_level.looping_sound)
+		wait = selected_level.looping_sound_interval
+		can_fire = TRUE
+	else
+		can_fire = FALSE
+
+	if(SSshuttle.emergency.mode == SHUTTLE_CALL || SSshuttle.emergency.mode == SHUTTLE_RECALL) // By god this is absolutely shit
+		old_shuttle_call_time_mod = 1 / old_shuttle_call_time_mod
+		SSshuttle.emergency.modTimer(old_shuttle_call_time_mod)
+		SSshuttle.emergency.modTimer(selected_level.shuttle_call_time_mod)
+
+	SEND_SIGNAL(src, COMSIG_SECURITY_LEVEL_CHANGED, selected_level.number_level)
 	SSnightshift.check_nightshift()
-	SSblackbox.record_feedback("tally", "security_level_changes", 1, get_security_level())
+	SSblackbox.record_feedback("tally", "security_level_changes", 1, selected_level.name)
+
+/**
+ * Handles announcements of the newly set security level
+ *
+ * Arguments:
+ * * selected_level The new security level that has been set
+ */
+/datum/controller/subsystem/security_level/proc/announce_security_level(datum/security_level/selected_level)
+	if(selected_level.number_level > current_security_level.number_level) // We are elevating to this level.
+		minor_announce(selected_level.elevating_to_announcemnt, "Attention! Security level elevated to [selected_level.name]:")
+	else // Going down
+		minor_announce(selected_level.lowering_to_announcement, "Attention! Security level lowered to [selected_level.name]:")
+	if(selected_level.sound)
+		sound_to_playing_players(selected_level.sound)
+
+/**
+ * Returns the current security level
+ *
+ * Arguments:
+ * * as_number Whether to return the security level as a text or as a number
```

Comment:
Your formatting on all of these is wrong, it should look like:

```suggestion
 * * as_number - Whether to return the security level as a text or as a number
```

Reply:
Got it